### ISSUE TITLE
Dynamic rules stop working after sheet re-attach

### ISSUE THREAD

**Issue by felthy** [cssinjs/jss#1187]
__Expected behavior:__
When my dynamic style changes, the new value should take effect.

__Describe the bug:__
When my dynamic style changes, the new rule is inserted into the sheet but there is another copy of the rule with the old value still in the sheet, and the old value is the one with precedence (the abandoned duplicate rule has a higher index in the sheet).

The other prerequisite for this bug to occur is that the stylesheet must have been detached then re-attached.

__Codesandbox link:__ 
https://codesandbox.io/s/jss-double-insert-zkhbq

The sandbox starts with a mounted component.
Click the button to change the color rule. This works.
Click the button to unmount the component, which detaches the sheet.
Click the button to mount the component again. This still looks correct, but there are now duplicate rules in the sheet.
Click the button to change the color rule. Nothing happens - now the old rule is taking precedence over the updated rule.

__Versions (please complete the following information):__
 - jss: `10.0.0-alpha.24`
 - Browser: `Chrome`
 - OS [e.g. Windows, macOS]: `macOS`
The bug is probably not dependent on browser, however.

**Comment by felthy** [comment]
The extra deploy causing the problem is at https://github.com/cssinjs/jss/blob/master/packages/jss/src/DomRenderer.js#L337:

```javascript
    // When rules are inserted using `insertRule` API, after `sheet.detach().attach()`
    // browsers remove those rules.
    // TODO figure out if its a bug and if it is known.
    // Workaround is to redeploy the sheet.
    if (this.hasInsertedRules) {
      this.hasInsertedRules = false
      this.deploy()
    }
```

**Comment by felthy** [comment]
I think the complicating factor might be this code at https://github.com/cssinjs/jss/blob/master/packages/jss/src/StyleSheet.js#L115:

```javascript
    // We can't add rules to a detached style node.
    // We will redeploy the sheet once user will attach it.
    this.deployed = false
```

The sequence is:

1. `addDynamicRules()` in `react-jss` reaches the above code that sets `Stylesheet.deployed` to `false`
2. `manageSheet()` in `react-jss` causes `StyleSheet.attach()` to be called
3. which calls `DomRenderer.attach()`
4. which because of `hasInsertedRules` calls `DomRenderer.deploy()`
5. which calls `DomRenderer.insertRules()` for the **first** time
6. ...Then, the `StyleSheet.attach()` call (same call in step 2) calls `StyleSheet.deploy()` because `Stylesheet.deployed` is `false` (from step 1)
7. which calls `DomRenderer.deploy()`
8. which calls `DomRenderer.insertRules()` for the **second** time

So I think the problem is we have two different special cases calling `DomRenderer.deploy()` for different reasons. I probably need your help here @kof but I'm thinking maybe the code above could be changed to something like this?

```javascript
    // We can't add rules to a detached style node.
    // We will redeploy the sheet once user will attach it.
    if (!this.renderer || !this.renderer.hasInsertedRules) {
      this.deployed = false
    }
```

It seems to fix it when I simulate the change in the debugger.

**Comment by felthy** [comment]
After some more debugging, I think the repeated call to `DomRenderer.deploy()` should be safe, but the problem is that during the second deploy, the browser's `CSSStyleSheet.insertRule(string, index)` is ignoring the index parameter. From the codesandbox, after the first call to deploy (on the second mount) the CSSStyleSheet contains:

0: `.Color-container-0-2-1 {}`
1: `.Color-container-1-0-2-3 { color: red; }`

Then the second deploy tries to do:
`c.insertRule('.Color-container-0-2-1 {}', 0)`
and
`c.insertRule('.Color-container-1-0-2-3 { color: red; }', 1)`

and the result is:

0: `.Color-container-0-2-1 {}`
1: `.Color-container-1-0-2-3 { color: red; }`
2: `.Color-container-0-2-1 {}`
3: `.Color-container-1-0-2-3 { color: red; }`

Looks like it could be a browser bug. 😭

**Comment by felthy** [comment]
On the initial mount, everything works fine because `this.hasInsertedRules` is false so there is only one call to `DomRenderer.deploy()`.

On subsequent re-attaches of the sheet, `DomRenderer.deploy()` gets called twice - the first time because `this.hasInsertedRules` is true, then again because `this.deployed` is false in the `StyleSheet`.

That latter happens because of the code at https://github.com/cssinjs/jss/blob/master/packages/jss/src/StyleSheet.js#L115:

```javascript
    // We can't add rules to a detached style node.
    // We will redeploy the sheet once user will attach it.
    this.deployed = false
```

...which was reached via the call to `addDynamicRules()` in the hook/hoc that occurs before the call to `manageSheet()`.

Is the problem just that these two things need to cooperate so that the deploy only happens once? The below seems to fix it for me (but I've only tried it via the debugger so far):

```javascript
    // We can't add rules to a detached style node.
    // We will redeploy the sheet once user will attach it.
    if (!(this.renderer && this.renderer. hasInsertedRules)) {
      this.deployed = false
    }
```

@kof you added the `this.hasInsertedRules` workaround and the add-rules-to-detached-sheets feature, do you think I'm on the right track here?

**Comment by kof** [comment]
You def. are on the right track!

**Comment by HenriBeck** [comment]
This could also be the reason for #1165

**Comment by felthy** [comment]
@HenriBeck happy for this issue to be closed in favour of #1165 if you prefer. My PR #1190 does solve the problem in that codesandbox too.

**Comment by kof** [comment]
released in 10.0.0-alpha.25
